src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs

```diff
@@ -59,36 +59,56 @@ public async Task<IEnumerable<EmissionsData>> GetCarbonIntensityAsync(IEnumerabl
         }
         return result;
     }
-    
+
     /// <inheritdoc />
-    public async Task<EmissionsForecast> GetCurrentCarbonIntensityForecastAsync(Location location)
+    public async Task<EmissionsForecast> GetCarbonIntensityForecastAsync(Location location, DateTimeOffset? generatedAt = null)
     {
-        this.Logger.LogInformation($"Getting carbon intensity forecast for location {location}");
+        // When generatedAt field null, we request the current forecast.
+        bool isCurrent = generatedAt == null;
+        this.Logger.LogInformation($"Getting carbon intensity forecast for location {location}" + (isCurrent ? "" : $"requested at {generatedAt}"));
 
         using (var activity = Activity.StartActivity())
         {
-            BalancingAuthority balancingAuthority = await this.GetBalancingAuthority(location, activity);
-            var data = await this.WattTimeClient.GetCurrentForecastAsync(balancingAuthority);
-
-            var duration = GetDurationFromGridEmissionDataPoints(data.ForecastData.FirstOrDefault(), data.ForecastData.Skip(1)?.FirstOrDefault());
-            
-            // Linq statement to convert WattTime forecast data into EmissionsData for the CarbonAware SDK.
-            var forecastData = data.ForecastData.Select(e => new EmissionsData() 
-            { 
-                Location = e.BalancingAuthorityAbbreviation, 
-                Rating = ConvertMoerToGramsPerKilowattHour(e.Value), 
-                Time = e.PointTime,
-                Duration = duration
-            });
-
-            return new EmissionsForecast()
+            var balancingAuthority = await this.GetBalancingAuthority(location, activity);
+            if (isCurrent)
+            {
+                var forecast = await this.WattTimeClient.GetCurrentForecastAsync(balancingAuthority); 
+                return ForecastToEmissionsForecast(forecast, location);
+            } 
+            else
             {
-                GeneratedAt = data.GeneratedAt,
-                Location = location,
-                ForecastData = forecastData,
-            };
+                var forecast = await this.WattTimeClient.GetForecastOnDateAsync(balancingAuthority, (DateTimeOffset)generatedAt!);
+                if (forecast == null)
+                {
+                    Exception ex = new ArgumentException($"No forecast was generated at the requested time {generatedAt}");
+                    activity?.SetStatus(ActivityStatusCode.Error, ex.Message);
+                    Logger.LogError(ex, ex.Message);
+                    throw ex;
+                }
+                return ForecastToEmissionsForecast(forecast, location);
+            }
         }
     }
+
+    private EmissionsForecast ForecastToEmissionsForecast(Forecast forecast, Location location) 
```

Comment:
Extracting helper that converts a Forecast+location into the EmissionsForecast object